src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsReadStream.cs

```diff
@@ -0,0 +1,340 @@
+
+        // The read pointer inside the buffer.
+        private int _readIndex = 0;
+
+        // The end of the data index in the buffer.
+        private int _readBufferDataEnd = 0;
+
+        /// <summary>
+        /// The number of bytes left in the packet to be consumed.
+        /// This is not necessarily the number of bytes available in the buffer.
+        /// </summary>
+        private int _packetDataLeft = 0;
+
+        /// <summary>
+        /// The number of bytes in the header, according to the header.
+        /// </summary>
+        private int _packetHeaderDataLength = 0;
+
+        private byte _packetHeaderType = 0;
+
+        private byte _packetStatus;
+
+        /// <summary>
+        /// Stored internally. It will be used for tracing.
+        /// </summary>
+        private int _spid = 0;
+
+        #endregion
+
+        #region Constructors
+
+        public TdsReadStream(Stream underlyingStream)
+        {
+            _readBuffer = new byte[TdsEnums.DEFAULT_LOGIN_PACKET_SIZE];
+            _underlyingStream = underlyingStream;
+        }
+
+        #endregion
+
+        #region Public Properties
+        /// <inheritdoc />
+        public override bool CanRead => _underlyingStream != null;
+
+        /// <inheritdoc />
+        public override bool CanSeek => false;
+
+        /// <inheritdoc />
+        public override bool CanWrite => false;
+
+        /// <inheritdoc />
+        public override long Length => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Position 
+        {
+            get => throw new NotImplementedException(); 
+            set => throw new NotImplementedException(); 
+        }
+
+        /// <inheritdoc />
+        public virtual int Spid => _spid;
+
+        #endregion
+
+        #region Public Methods
+
+        /// <inheritdoc />
+        public override async ValueTask DisposeAsync()
+        {
+            await _underlyingStream.DisposeAsync();
+            _underlyingStream = null;
+            _readBuffer = null;
+        }
+
+        /// <inheritdoc />
+        public override void Flush() => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public async ValueTask<byte> PeekByteAsync(bool isAsync, CancellationToken ct)
+        {
+            // If we have logically finished reading the packet, or if we have 
+            // reached the end of the buffer, then we need to position the buffer at the beginning of next
+            // packet start.
+            await PrepareBufferIfNeeded(isAsync, ct).ConfigureAwait(false);
+            return _readBuffer[_readIndex];
+        }
+
+        /// <inheritdoc />
+        public override int Read(Span<byte> buffer)
+        {
+            int lengthToFill = buffer.Length;
+            int totalRead = 0;
+            while (lengthToFill > 0)
+            {
+                PrepareBufferIfNeeded(isAsync: false, CancellationToken.None).ConfigureAwait(false);
```

Comment:
I think you meant to await this or use the result somehow? This one is giving me the following message https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2012

Reply:
I called the method with isAsync -> False.
When it returns, I know that I dont have to wait, and it would have modified all the states of this class when it returns.

Reply:
I see what you're saying and generally speaking I agree, but I think the code analysis rule explanation does make sense and is a better general rule to follow. If it completes synchronously (which it's supposed to), would it hurt to await it (I guess in this case via .GetResult())?

Reply:
I took a slightly different route. 
I consumed the VT in an assert.